### PR TITLE
Add generic HDF5 dataset

### LINKED ISSUE #1026: Add catalog dataset for HDF formats not supported by pandas
## Description
HDF files which were not created by `pandas.to_hdf` sometimes cannot be read using `pandas.read_hdf`. Since Kedro's HDFDataSet depends on pandas, these files cannot be added to the catalog all.

## Context
Currently, the dynamic simulation software employed in my research group outputs exclusively `.h5` files which contain information we wish to feed to ML models using Kedro. Currently we use a Python script which converts these HDF files into CSVs so we can track using Kedro, yet this is an inefficient process as we are required to rewrite thousands of files to process them.

Ideally, we would like to add out dataset to our data catalog just like we do with our CSV datasets, but in a way that was able to read any kind of HDF file, unlike `kedro.extras.datasets.pandas.HDFDataSet`.

Given pandas cannot read HDF files which do not conform to its specs (apparently by design according [to this issue](https://github.com/pandas-dev/pandas/issues/9539)), this simple addition would benefit any user who may store information in HDF files, be it because it is their preferred storage method or (like in our case) they use some software which directly outputs HDF. 

## Possible Implementation
My research colleague @Eric-OG and I believe we can implement this on our own. I think it's worth noting *it'd be our first contribution to an open source project*, but we've read the guidelines and so forth.

 It would basically involve copying one of the existing datasets (possible even `pandas.HDFDataSet`) and adapting it to use another library. We planned on using [`h5py`](https://www.h5py.org/).

## Possible Alternatives
* Performing part of our data processing pipelines without Kedro; this is cumbersome and can get harder to maintain, especially since our code will likely be used by new researchers next year;
* Converting the files to another type already implemented; this is what we do today but it's simply inefficient.

### PR DESCRIPTION
**Pull request by tomaz-suller**
## Description
Add dataset and tests for HDF5 files which may not be supported by pandas. Resolves #1026.

pandas only supports HDF5 files with tabular data, and then only those written by pandas; our use case involved reading files generated by another program, so we implemented a data set which uses h5py. 

## Development notes
We implemented a new dataset (`hdf5.H5pyDataSet`) based on the existing `pandas.HDFDataSet` but using h5py and its low-level API to allow binary IO required by fsspec.

We used our implementation to read (though never write) HDF5 files successfully in our research group. We also developed unit tests (again based on those for the pandas data set) to reach full coverage according to the guidelines.

## Checklist

- [X] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [X] Updated the documentation to reflect the code changes
- [X] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [X] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1964"></a>

### CODE REVIEW

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (merelcht) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:51]: We've started to remove this example in other datasets (https://github.com/kedro-org/kedro/pull/1962) because it doesn't add anything, so let's leave it out here as well.

File: tests/extras/datasets/hdf5/test_h5py_dataset.py
Comment (merelcht) [inline tests/extras/datasets/hdf5/test_h5py_dataset.py:23]: Perhaps name this `h5py_data_set` to differentiate more from the existing hdf dataset.

File: tests/extras/datasets/hdf5/test_h5py_dataset.py
Comment (merelcht) [inline tests/extras/datasets/hdf5/test_h5py_dataset.py:34]: Same here.

**Review by merelcht — APPROVED** [review]
Thanks for this contribution @tomaz-suller I don't know much about HDF5 so I'll add some other reviewers, but the code looks good to me 👍 

In terms of the DCO check, have you tried following the instructions the check suggests:
```
To add your Signed-off-by line to every commit in this branch:

1. Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
2. In your local branch, run: git rebase HEAD~3 --signoff
3. Force push your changes to overwrite the branch: git push --force-with-lease origin master
```

If that didn't work, let me know, and I'll see if I can get it to pass 🙂

**Comment by merelcht** [comment]
Actually one more question, is `h5py` a requirement for this dataset? If so, you'll have to add it to [`setup.py`](https://github.com/kedro-org/kedro/blob/main/setup.py) just like the other datasets.

File: tests/extras/datasets/hdf5/test_h5py_dataset.py
Comment (noklam) [inline tests/extras/datasets/hdf5/test_h5py_dataset.py:65]: Can we remove this function and simplified this as

```
assert isinstance(expected, h5py.Dataset) and isinstance(
    received, h5py.Dataset
)
np.testing.assert_equal(expected[()], received[()])
```

The function assumes both object need to be h5py.DataSet otherwise it will fail, so there is no need to do type checking.
In addition the boolean value isn't used in the following test so this seems not necessary.

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (noklam) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:67]: Can you explain why is it not thread-safe otherwise? What makes `h5py.HDF5DataSet` more special than the others?

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (noklam) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:80]: I don't know h5py well enough, are these arguments never change? would there be need that users need to change these settings/flags?

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (noklam) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:100]:
```suggestion
    def __h5py_to_binary(h5f: h5py.File, save_args):
        bio = BytesIO()
        with h5py.File(bio, "w", **save_args) as f:
            for _, value in h5f.items():
                h5f.copy(
                    value,
                    f,
                    expand_soft=True,
                    expand_external=True,
                    expand_refs=True,
                )
            f.close()
            return bio.getvalue()
```

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (noklam) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:72]: Can these be just `_` instead of `__`?  also would be good to add the type for `binary`

File: kedro/extras/datasets/hdf5/h5py_dataset.py
Comment (noklam) [inline kedro/extras/datasets/hdf5/h5py_dataset.py:186]: Why is this guaranteed binary_data?

**Review by noklam — COMMENTED** [review]
thanks for the PR! I have left some comments for the 1st draft.

**Comment by merelcht** [comment]
Hi @tomaz-suller do you still want to complete this PR? We'd like to get all PRs related to datasets to be merged soon now we're moving our datasets code to a different package (see our [medium blog post](https://medium.com/quantumblack/keeping-up-with-kedro-the-latest-developments-in-our-development-workflow-framework-cbcc415eea9c) for more details)

Otherwise we'll close this PR and ask you to re-open it on the new repo when it's ready.

**Comment by tomaz-suller** [comment]
Hi @merelcht. I'm getting back to this next Saturday, so I'd say it'll still be a week or two before the code is ready for a merge. 

I'm fine with opening the PR to the new repo so I don't get in the way.

**Comment by merelcht** [comment]
Okay, no worries at all @tomaz-suller ! Let me or anyone from the maintainer team know if you need help 🙂 We really appreciate your contribution!

**Pull request closed without merging**
